kernel/src/fs/ramfs/fs.rs

```diff
@@ -351,29 +356,46 @@ impl DirEntry {
         } else if name == ".." {
             Some((1, self.parent.upgrade().unwrap()))
         } else {
-            self.children
-                .idxes_and_items()
-                .find(|(_, (child, _))| child.as_str().unwrap() == name)
-                .map(|(idx, (_, inode))| (idx + 2, inode.clone()))
+            let idx = *self.idx_map.get(name.as_bytes())?;
+            let target_inode = self
+                .children
+                .get(idx)
+                .map(|(n, inode)| {
+                    debug_assert_eq!(name, n.as_str().unwrap());
+                    inode.clone()
+                })
+                .unwrap();
+            Some((idx + NUM_SPECIAL_ENTRIES, target_inode))
         }
     }
 
     fn append_entry(&mut self, name: &str, inode: Arc<RamInode>) -> usize {
-        self.children.put((CStr256::from(name), inode))
+        let name = CStr256::from(name);
+        let idx = self.children.put((name, inode));
+        self.idx_map.insert(name, idx);
+        idx
     }
 
     fn remove_entry(&mut self, idx: usize) -> Option<(CStr256, Arc<RamInode>)> {
-        assert!(idx >= 2);
-        self.children.remove(idx - 2)
+        assert!(idx >= NUM_SPECIAL_ENTRIES);
+        let removed = self.children.remove(idx - NUM_SPECIAL_ENTRIES)?;
+        self.idx_map.remove(&removed.0);
+        Some(removed)
     }
 
     fn substitute_entry(
         &mut self,
         idx: usize,
         new_entry: (CStr256, Arc<RamInode>),
     ) -> Option<(CStr256, Arc<RamInode>)> {
-        assert!(idx >= 2);
-        self.children.put_at(idx - 2, new_entry)
+        assert!(idx >= NUM_SPECIAL_ENTRIES);
+        let name = new_entry.0;
+        let idx_children = idx - NUM_SPECIAL_ENTRIES;
+
+        let substitute = self.children.put_at(idx_children, new_entry)?;
+        self.idx_map.remove(&substitute.0);
+        self.idx_map.insert(name, idx_children);
```

Comment:
The call to `remove` is unnecessary.  The subsequent call to `insert` will replace and return any old entry if it exists.

Reply:
I'm afraid not, the `idx_map` now caches mapping "name -> idx", the `substitute_entry()` is used in `rename()`. In this case the `idx` is the same but the name is not, the old name and the new name appear in different KVs in `idx_map`, so we need to remove the old one then insert the new one.